promql/promqltest/test.go

```diff
@@ -693,29 +816,79 @@ func (ev *evalCmd) expectMetric(pos int, m labels.Labels, vals ...parser.Sequenc
 	ev.expected[h] = entry{pos: pos, vals: vals}
 }
 
+// validateExpectedAnnotations validates expected messages and regex match actual annotations.
+func validateExpectedAnnotations(expr string, expectedAnnotations []expectCmd, actualAnnotations []string, line int, annotationType string, annos annotations.Annotations) error {
+	if len(expectedAnnotations) == 0 {
+		return nil
+	}
+
+	if len(actualAnnotations) == 0 {
+		return fmt.Errorf("expected %s annotations but none were found for query %q (line %d)", annotationType, expr, line)
+	}
+
+	// Check if all expected annotations are found in actual.
+	for _, e := range expectedAnnotations {
+		matchFound := slices.ContainsFunc(actualAnnotations, e.CheckMatch)
+		if !matchFound {
+			return fmt.Errorf(`expected %s annotation matching %s %q but no matching annotation was found for query %q (line %d), found: %v`, annotationType, e.Type(), e.String(), expr, line, annos.AsErrors())
+		}
+	}
+
+	// Check if all actual annotations have a corresponding expected annotation.
+	for _, anno := range actualAnnotations {
+		matchFound := slices.ContainsFunc(expectedAnnotations, func(e expectCmd) bool {
+			return e.CheckMatch(anno)
+		})
+		if !matchFound {
+			return fmt.Errorf("unexpected %s annotation %q found for query %s (line %d)", annotationType, anno, expr, line)
+		}
+	}
+
+	return nil
+}
+
 // checkAnnotations asserts if the annotations match the expectations.
 func (ev *evalCmd) checkAnnotations(expr string, annos annotations.Annotations) error {
 	countWarnings, countInfo := annos.CountWarningsAndInfo()
 	switch {
-	case ev.ordered:
-		// Ignore annotations if testing for order.
-	case !ev.warn && countWarnings > 0:
-		return fmt.Errorf("unexpected warnings evaluating query %q (line %d): %v", expr, ev.line, annos.AsErrors())
 	case ev.warn && countWarnings == 0:
 		return fmt.Errorf("expected warnings evaluating query %q (line %d) but got none", expr, ev.line)
-	case !ev.info && countInfo > 0:
-		return fmt.Errorf("unexpected info annotations evaluating query %q (line %d): %v", expr, ev.line, annos.AsErrors())
 	case ev.info && countInfo == 0:
 		return fmt.Errorf("expected info annotations evaluating query %q (line %d) but got none", expr, ev.line)
 	}
+
+	var warnings, infos []string
+
+	for _, err := range annos {
+		switch {
+		case errors.Is(err, annotations.PromQLWarning):
+			warnings = append(warnings, err.Error())
+		case errors.Is(err, annotations.PromQLInfo):
+			infos = append(infos, err.Error())
+		default:
+			return errors.New("unexpected annotation type, must be either info or warn")
```

Comment:
Could you please include the unexpected annotation here? Otherwise it will be difficult to understand what went wrong.

Reply:
Done!